### PR TITLE
Adding container for packet capture api to tigera manager

### PR DESCRIPTION
**Pull request by asincu**
## Description

PacketCapture Download API exposes the API to retrieve pcap files generated by a PacketCapture. This enables personas
like DevOps or Platform Engineers to retrieve files backed by a namespace RBAC.  

PacketCapture API is a service that is deployed as container part of tigera-manager deployment. By default, it is
installed by Tigera operator. This service will only be deployed only for Standalone and Management clusters, and requests
will be routed to the appropriate managed clusters.

Voltron container has route defined for any requests that starts with `/packet-capture/` prefix. It will strip away that 
prefix and forward it to the container where this service resides.

## Configuration and permissions

| ENV        | Default value          | Description  |
| ------------- |:-------------:| -----:|
| PACKETCAPTURE_API_PORT      | `8444` | Local Port to start the service |
| PACKETCAPTURE_API_HOST      | <empty>, defaults to localhost      |   Host for the service |
| PACKETCAPTURE_API_LOG_LEVEL | `Info`      |    Log Level across service |
| PACKETCAPTURE_API_DEX_ENABLED | `False`      |    Enable Dex for authentication |
| PACKETCAPTURE_API_DEX_ISSUER | `https://127.0.0.1:5556/dex`      |    Dex Setup Configuration |
| PACKETCAPTURE_API_DEX_CLIENT_ID | `tigera-manager`      |    Dex Setup Configuration |
| PACKETCAPTURE_API_DEX_JWK_URL | `https://tigera-dex.tigera-dex.svc.cluster.local:5556/dex/keys`     |    Dex Setup Configuration |
| PACKETCAPTURE_API_DEX_USER_CLAIM | `email`      |    Dex Setup Configuration |
| PACKETCAPTURE_API_DEX_GROUPS_CLAIM | <empty>      |    Dex Setup Configuration |
| PACKETCAPTURE_API_DEX_USERNAME_PREFIX | <empty>      |    Dex Setup Configuration |
| PACKETCAPTURE_API_DEX_GROUPS_PREFIX | <empty>      |    Dex Setup Configuration |
| PACKETCAPTURE_API_MULTI_CLUSTER_FORWARDING_CA | `/manager-tls/cert`      |    CA certificate for multicluster communication |
| PACKETCAPTURE_API_MULTI_CLUSTER_FORWARDING_ENDPOINT | `https://localhost:9443`      |    CA endpoint for multicluster communication |

This API makes use of tigera-manager service account and requires the following permissions:
- GET for api group `projectcalico.org` for resource `packetcaptures`
- CREATE for api group `projectcalico.org` for resource `authenticationreviews`
- LIST,WATCH,GET for api group `projectcalico.org` for resource `managedclusters`
- LIST for core v1 group for resource `pods`
- CREATE for core v1 group for subresource `pods/exec`
- CREATE for api group `authorization.k8s.io` for resource `subjectaccessreviews`

## For PR author

- [ ] Tests for change.
- [ ] If changing pkg/apis/, run `make gen-files`
- [ ] If changing versions, run `make gen-versions`

## For PR reviewers

A note for code reviewers - all pull requests must have the following:

- [ ] Milestone set according to targeted release.
- [ ] Appropriate labels:
  - `kind/bug` if this is a bugfix.
  - `kind/enhancement` if this is a a new feature.
  - `enterprise` if this PR applies to Calico Enterprise only.

### CODE REVIEW

File: pkg/render/manager.go
Comment (rene-dekker) [inline pkg/render/manager.go:481]: Maybe rename to managerPacketCaptureLivenessProbe

File: pkg/render/manager.go
Comment (rene-dekker) [inline pkg/render/manager.go:315]: I was wondering if we should start adding resources and limits to the manager deployment. I'm not sure if you would be able to come up with good numbers or otherwise could add a comment for this container detailing how much it would roughly require going forward to your best estimate.

File: pkg/render/manager.go
Comment (rene-dekker) [inline pkg/render/manager.go:588]: unnessecary  string(...)

**Review by rene-dekker — APPROVED** [review]
lgtm

File: pkg/render/manager.go
Comment (caseydavenport) [inline pkg/render/manager.go:599]: Is there not a more specific name we could give this? like `tigera-packetcapture-server` or something that indicates it's purpose?

The container itself, I assume, is not a packet capture :D

File: pkg/render/manager.go
Comment (caseydavenport) [inline pkg/render/manager.go:609]: Where did these resource requirement values come from? Are they for the entire manager or just the packetcapture component?

The function name implies the entire manager should receive these.

File: pkg/render/manager.go
Comment (caseydavenport) [inline pkg/render/manager.go:779]: Is it possible to narrow this permission at all? Can we at least limit to exec only for calico/node pods?

**Review by caseydavenport — CHANGES_REQUESTED** [review]

File: pkg/render/manager.go
Comment (asincu) [reply pkg/render/manager.go:609]: They are being applied for a single container.

File: pkg/render/manager.go
Comment (asincu) [reply pkg/render/manager.go:779]: Sure, I will give it a try. I think it should be fluentd pods.

File: pkg/render/manager.go
Comment (asincu) [reply pkg/render/manager.go:492]: These are the default delays we use for all our probes.

File: pkg/render/manager.go
Comment (asincu) [reply pkg/render/manager.go:488]: The probes are accessed via port 9443. This is the only one open for the pod.

Comment (rene-dekker) [reply pkg/render/manager.go:488]: Voltron is listening on this port and doing backend proxy magic.

File: pkg/render/manager.go
Comment (caseydavenport) [reply pkg/render/manager.go:609]: we should probably rename the function to indicate that then. 

I'm also interested in how you came to these numbers.

Comment (asincu) [reply pkg/render/manager.go:609]: I simply added the values that were in the K8s example. I do not have a very scientific approach to this.

File: pkg/render/manager.go
Comment (asincu) [reply pkg/render/manager.go:781]: @caseydavenport I think I need to create a role and limit this to a particular namespace. I think ResourceName might require the full name of the pod.

File: pkg/render/manager.go
Comment (matthewdupre) [reply pkg/render/manager.go:492]: It would be fine for us to reduce this if we don't have a reason to expect the health check to take a long time to get going (I'm _guessing_ that this one becomes healthy very quickly, so maybe 10s would be fine), but since the manager pod is a singleton I think the impact of this time not being precisely calibrated is almost nothing.

As such I suggest not interfering with the 3.8 schedule for this, we can always make a follow-up PR if you feel strongly.

File: pkg/render/manager.go
Comment (matthewdupre) [reply pkg/render/manager.go:765]: Can we do this as a follow up?  Delaying 3.8 testing further is problematic.

**Comment by matthewdupre** [comment]
@asincu the build is failing, I think because some component version based autogeneration hasn't been ran.

File: pkg/render/manager.go
Comment (asincu) [reply pkg/render/manager.go:492]: Talked offline. We are not changing the default delays now.

File: pkg/render/manager.go
Comment (asincu) [reply pkg/render/manager.go]: Talked offline: Removing the resource requirements.

File: pkg/render/fluentd.go
Comment (caseydavenport) [inline pkg/render/fluentd.go:365]: Definitely need a comment explaining why this pod/exec role is here and what it is used for.

File: pkg/render/manager.go
Comment (caseydavenport) [reply pkg/render/manager.go:765]: @asincu are you planning to remove this permission here now that we are granting it via a Role / RoleBinding?

**Review by caseydavenport — CHANGES_REQUESTED** [review]
Just two comments - one regarding a missing comment, the other about duplicate permission granting.

**Comment by asincu** [comment]
@caseydavenport Do you mind checking the last 3 commits again ?

**Review by caseydavenport — APPROVED** [review]
LGTM - Make sure to squash!

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and delete the branch after I've merged it.

**Comment by marvin-tigera** [comment]
I'm sorry but I failed delete the branch after merging the pull request.

**Pull request merged**
